lib/que/worker.rb

```diff
@@ -1,184 +1,96 @@
-    include MonitorMixin
-
-    attr_reader :thread, :state, :queue
+    # Defines the time a worker will wait before checking Postgres for its next job
+    WAKE_INTERVAL = 5
+    DEFAULT_QUEUE = ''
+    JOB_INSTANCE_FIELDS = %i[queue priority run_at job_id].freeze
 
-    def initialize(queue = '')
-      super() # For MonitorMixin.
+    def initialize(queue: DEFAULT_QUEUE, wake_interval: WAKE_INTERVAL)
       @queue  = queue
-      @state  = :working
-      @thread = Thread.new { work_loop }
-      @thread.abort_on_exception = true
+      @wake_interval = wake_interval
+      @stop = false
     end
 
-    def alive?
-      !!@thread.status
-    end
+    def work_loop
+      return if @stop
 
-    def sleeping?
-      synchronize { _sleeping? }
-    end
+      loop do
+        case work
+        when :job_not_found, :error then sleep(@wake_interval)
+        when :job_worked then nil # immediately find a new job to work
+        end
 
-    def working?
-      synchronize { @state == :working }
+        break if @stop
+      end
     end
 
-    def wake!
-      synchronize do
-        if sleeping?
-          # Have to set the state here so that another thread checking
-          # immediately after this won't see the worker as asleep.
-          @state = :working
-          @thread.wakeup
-          true
+    def work
+      Que.adapter.checkout do
+        with_locked_job do |job|
+          return :job_not_found if job.nil?
+
+          # Check that the job hasn't just been worked by another worker (it's possible to
+          # lock a job that's just been destroyed because pg locks don't obey MVCC). If it
+          # has been worked, act as if we've worked it.
+          return :job_worked unless job_exists?(job)
+
+          begin
+            klass = class_for(job[:job_class])
+            # TODO: _run -> run_and_destroy(*job[:args])
+            klass.new(job)._run
+          rescue => error
+            # For compatibility with que-failure, we need to allow failure handlers to be
+            # defined on the job class.
+            if klass.respond_to?(:handle_job_failure)
+              klass.handle_job_failure(error, job)
+            else
+              handle_job_failure(error, job)
+            end
+          end
+          :job_worked
         end
       end
+    rescue PG::Error => _error
+      # In the event that our Postgres connection is bad, we don't want that error to halt
+      # the work loop. Instead, we should let the work loop sleep and retry.
+      :error
     end
 
-    # This needs to be called when trapping a signal, so it can't lock the monitor.
-    def stop
+    def stop!
       @stop = true
-      @thread.wakeup if _sleeping?
-    end
-
-    def wait_until_stopped
-      wait while alive?
     end
 
     private
 
-    # Sleep very briefly while waiting for a thread to get somewhere.
-    def wait
-      sleep 0.0001
+    # Set the error and retry with back-off
+    def handle_job_failure(error, job)
```

Comment:
Just want to clarify something here. If we don't use any failure strategies (from que-failure) we will retry forever with back off as we haven't overridden `handle_job_failure`.

Reply:
This is fine I think, as we force people to specify the failure strategy in PS and exponential backoff is a legit default behaviour.

Reply:
What we should do is add jitter, which I'll do now.